src/distPlus/symbolic/SymbolicDist.re

```diff
@@ -296,103 +313,407 @@ module GenericSimple = {
   };
 };
 
-module PointwiseAddDistributionsWeighted = {
-  type t = pointwiseAdd;
+module DistTree = {
+  type nodeResult = [
+    | `Simple(dist)
+    // RenderedShape: continuous xyShape, discrete xyShape, total value.
+    | `RenderedShape(DistTypes.continuousShape, DistTypes.discreteShape, integral)
+  ];
 
-  let normalizeWeights = (dists: t) => {
-    let total = dists |> E.A.fmap(snd) |> E.A.Floats.sum;
-    dists |> E.A.fmap(((a, b)) => (a, b /. total));
+  let evaluateDistribution = (d: dist): nodeResult => {
+    `Simple(d)
   };
 
-  let pdf = (x: float, dists: t) =>
-    dists
-    |> E.A.fmap(((e, w)) => GenericSimple.pdf(x, e) *. w)
-    |> E.A.Floats.sum;
-
-  let min = (dists: t) =>
-    dists |> E.A.fmap(d => d |> fst |> GenericSimple.min) |> E.A.min;
-
-  let max = (dists: t) =>
-    dists |> E.A.fmap(d => d |> fst |> GenericSimple.max) |> E.A.max;
-
-  let discreteShape = (dists: t, sampleCount: int) => {
-    let discrete =
-      dists
-      |> E.A.fmap(((r, e)) =>
-           r
-           |> (
-             fun
-             | `Float(r) => Some((r, e))
-             | _ => None
-           )
-         )
-      |> E.A.O.concatSomes
-      |> E.A.fmap(((x, y)) =>
-           ({xs: [|x|], ys: [|y|]}: DistTypes.xyShape)
-         )
-      |> Distributions.Discrete.reduce((+.));
-    discrete;
+  // This is a performance bottleneck!
+  // Using raw JS here so we can use native for loops and access array elements
+  // directly, without option checks.
```

Comment:
Kudos for noticing and doing this to resolve. I have found that I was able to improve the ReasonML performance somewhat, though I imagine it would be impossible to beat raw js.

Reply:
See this function in XYShape.re:

```reason
  let integrateWithTriangles = ({xs, ys}) => {
    let length = E.A.length(xs);
    let cumulativeY = Belt.Array.make(length, 0.0);
    for (x in 0 to E.A.length(xs) - 2) {
      let _ =
        Belt.Array.set(
          cumulativeY,
          x + 1,
          (xs[x + 1] -. xs[x])
          *. ((ys[x] +. ys[x + 1]) /. 2.)
          +. cumulativeY[x],
        );
      ();
    };
    Some({xs, ys: cumulativeY});
  };
```

I think this is similar to what you are doing, and seemed reasonably performant. I'm not exactly sure what you are referring to option checks in particular, as the types aren't optionals?

Reply:
Oh ...

I had operated under the assumption that array access only works via `Belt.Array.get`. LOL. Alright then, I think I can Reason-ify it pretty easily then

Reply:
No worries, the Belt api takes some time to get used to. 

When you Reason-ify it, I suggest trying to move this to XYShape or Distributions. We have some similar functionality in XYShape, it's possible you could reuse some of that or improve on that then call it here. 

I'm not actually sure exactly what this does, so can't say for sure how it would integrate with other parts.